partitura/musicanalysis/note_array_to_score.py

```diff
@@ -0,0 +1,359 @@
+    This function requires a note array that contains time signatures and key signatures(optional - can also estimate it automatically).
+    Note array should contain the following fields:
+    - onset_div or onset_beat
+    - duration_div or duration_beat
+    - pitch
+    - ts_beats
+    - ts_beat_type
+    - key_mode(optional)
+    - key_fifths(optional)
+    - id(required but can also be empty)
+
+    Parameters
+    ----------
+    note_array : structure array or list of structured arrays.
+        A note array with the following fields:
+        - onset_div or onset_beat
+        - duration_div or duration_beat
+        - pitch
+        - ts_beats
+        - ts_beat_type
+        - key_mode(optional)
+        - key_fifths(optional)
+        - id(required but can also be empty)
+    divs : int (optional)
+        Divs is optional when note_array contains onset_beat and duration_beat. Otherwise it is required.
+    key_sigs : list (optional)
+        List of key signatures. Each key signature is a list of [onset_div, key_name, end_div].
+    assign_note_ids : bool (optional)
+        Assign note_ids.
+    estimate_key: bool (optional)
+        estimate key from note_array.
+    sanitize: bool (optional)
+        sanitize the part by adding measures, tying notes, and finding tuplets.
+    return_part: bool (optional)
+        Return a Partitura score object instead of a part.
+
+    Returns
+    -------
+    part : Score or Part
+        A part or score object.
+    """
+    if isinstance(note_array, list):
+        parts = [
+            note_array_to_score(note_array=x, name_id=str(i), assign_note_ids=assign_note_ids,
+                                return_part=True, divs=divs, estimate_key=estimate_key, sanitize=sanitize,
+                                part_name=name_id+"_P"+str(i)) for
+            i, x in enumerate(note_array)]
+        return score.Score(partlist=parts)
+
+    if not isinstance(note_array, np.ndarray):
+        raise TypeError("The note array does not have the correct format.")
+
+    if len(note_array) == 0:
+        raise ValueError("The note array is empty.")
+
+    # Test Note array for negative durations
+    if "duration_div" in note_array.dtype.names:
+        assert np.all(note_array["duration_div"] >= 0), "Note array contains negative durations."
+    elif "duration_beat" in note_array.dtype.names:
+        assert np.all(note_array["duration_beat"] >= 0), "Note array contains negative durations."
+
+    # Note id creation or re-assignment
+    if "id" not in note_array.dtype.names:
+        note_ids = ["{}n{:4d}".format(name_id, i) for i in range(len(note_array))]
+        note_array = rfn.append_fields(note_array, "id", np.array(note_ids, dtype='<U256'))
+    elif assign_note_ids or np.all(note_array["id"] == note_array["id"][0]):
+        note_ids = ["{}n{:4d}".format(name_id, i) for i in range(len(note_array))]
+        note_array["id"] = np.array(note_ids)
+
+    dtypes = note_array.dtype.names
```

Comment:
here dtypes is a variable but the same info is used in lines 224, 226, 230

Reply:
Not, quite the same they might be updated with the ids.